internal/controller/deletion/deletion.go

```diff
@@ -0,0 +1,256 @@
+			if deleted || (err != nil && cierrors.IsNotOwnedObject(err)) {
+				ctx.Logger.Sugar().Debugf(
+					"Removing object (%s) from ClusterInstance.Status.ManifestsRendered",
+					resourceId)
+				manifestsRendered = append(manifestsRendered[:manifestRefIndex], manifestsRendered[manifestRefIndex+1:]...)
+				continue
+			}
+
+			if err != nil {
+				ctx.Logger.Sugar().Errorf("unable to delete object (%s)", resourceId, zap.Error(err))
+				errs = append(errs, err)
+				if cierrors.IsDeletionTimeoutError(err) {
+					manifestRef.UpdateStatus(v1alpha1.ManifestDeletionTimedOut, err.Error())
+					return
+				}
+				manifestRef.UpdateStatus(v1alpha1.ManifestDeletionFailure, err.Error())
+				continue
+			}
+
+			if manifestRef.Status != v1alpha1.ManifestDeletionInProgress {
+				ctx.Logger.Sugar().Debugf(
+					"Updating object (%s) Status from from (%s) to (%s)",
+					resourceId, manifestRef.Status, v1alpha1.ManifestDeletionInProgress)
+				manifestRef.UpdateStatus(v1alpha1.ManifestDeletionInProgress, "")
+			}
+
+			syncWaveGroupDeleted = false
+		}
+
+		if len(errs) > 0 || !syncWaveGroupDeleted {
+			return
+		}
+	}
+
+	ctx.Logger.Info("Successfully deleted objects")
+	deletionCompleted = true
+	return
+}
+
+// DeleteObject ensures the deletion of an owned object.
+func (ctx *DeletionContext) DeleteObject(object ci.RenderedObject, ownerRefLabel string) (bool, error) {
+	obj := ptr.To(object.GetObject())
+
+	resourceId := object.GetResourceId()
+
+	if exists, err := ctx.ensureObjectExists(obj); !exists && err == nil {
+		return true, nil
+	} else if err != nil {
+		return false, err
+	}
+
+	if err := ci.VerifyOwnership(obj, ownerRefLabel); err != nil {
+		ctx.Logger.Sugar().Errorf("Ownership verification failed for object (%s): %v", resourceId, err.Error())
+		return false, err
+	}
+
+	if obj.GetDeletionTimestamp().IsZero() {
+		if err := ctx.initiateObjectDeletion(obj); err != nil {
+			return false, err
+		}
+	} else if timeoutExceeded(obj.GetDeletionTimestamp().Time, ctx.TimeoutDuration) {
+		return false, cierrors.NewDeletionTimeoutError(resourceId)
+	}
+
+	ctx.Logger.Sugar().Infof("Waiting for object (%s) to be deleted", resourceId)
+	return false, nil
+}
+
+// Utility functions
+
+func (ctx *DeletionContext) buildDeletionObjects(clusterInstance *v1alpha1.ClusterInstance) []ci.RenderedObject {
+	objects := make([]ci.RenderedObject, 0, len(clusterInstance.Status.ManifestsRendered))
+	for _, manifest := range clusterInstance.Status.ManifestsRendered {
+		object := ci.RenderedObject{}
+		if err := object.SetObject(map[string]interface{}{
+			"apiVersion": *manifest.APIGroup,
+			"kind":       manifest.Kind,
+			"metadata": map[string]interface{}{
+				"name":      manifest.Name,
+				"namespace": manifest.Namespace,
+			},
+		}); err == nil {
+			objects = append(objects, object)
+		}
+	}
+	return objects
+}
+
+func (ctx *DeletionContext) ensureObjectExists(obj client.Object) (bool, error) {
```

Comment:
This function isn't "ensuring" anything. I think it would be better as `objectExists` or `doesObjectExist`

Reply:
Or maybe `isObjectDeleted` would be better. Then you don't have to reverse the logic in the caller.

Reply:
I had considered something like `isObjectDeleted` previously, however, since we are merely checking for the existence of an object - it felt counter-intuitive. However, since the absence of the object allows us to conclude that the object must have been deleted (if previously created) - I think it's okay to rename it to`isObjectDeleted`.